topology/src/test/java/io/camunda/zeebe/topology/api/TopologyManagementRequestsHandlerTest.java

```diff
@@ -0,0 +1,118 @@
+  private final TopologyManagementRequestsHandler handler =
+      new TopologyManagementRequestsHandler(coordinator, new TestConcurrencyControl());
+
+  @ParameterizedTest
+  @MethodSource("provideReassignmentParameters")
+  void shouldReassignPartitionsRoundRobin(
+      final int partitionCount,
+      final int replicationFactor,
+      final int oldClusterSize,
+      final int newClusterSize) {
+    // given
+    final var expectedNewDistribution =
+        new RoundRobinPartitionDistributor()
+            .distributePartitions(
+                getClusterMembers(newClusterSize),
+                getSortedPartitionIds(partitionCount),
+                replicationFactor);
+
+    final var oldDistribution =
+        new RoundRobinPartitionDistributor()
+            .distributePartitions(
+                getClusterMembers(oldClusterSize),
+                getSortedPartitionIds(partitionCount),
+                replicationFactor);
+    var oldClusterTopology = TopologyUtil.getClusterTopologyFrom(oldDistribution);
+    for (int i = oldClusterSize; i < newClusterSize; i++) {
+      oldClusterTopology =
+          oldClusterTopology.addMember(
+              MemberId.from(Integer.toString(i)), MemberState.initializeAsActive(Map.of()));
+    }
+    coordinator.setCurrentTopology(oldClusterTopology);
+
+    // when
+    handler
+        .reassignPartitions(new ReassignPartitionsRequest(getClusterMembers(newClusterSize)))
+        .join();
+
+    // apply operations to generate new topology
+    final var topologyChangeSimulator =
+        new TopologyChangeAppliersImpl(
+            new NoopPartitionChangeExecutor(), new NoopTopologyMembershipChangeExecutor());
+    var newTopology = oldClusterTopology.startTopologyChange(coordinator.getLastAppliedOperation());
+
+    while (newTopology.hasPendingChanges()) {
+      final var operation = newTopology.changes().pendingOperations().get(0);
+      final var applier = topologyChangeSimulator.getApplier(operation);
+      newTopology = newTopology.updateMember(operation.memberId(), applier.init(newTopology).get());
+      newTopology = newTopology.advanceTopologyChange(operation.memberId(), applier.apply().join());
+    }
+
+    // then
+    final var newDistribution = TopologyUtil.getPartitionDistributionFrom(newTopology, "temp");
+    assertThat(newDistribution).isEqualTo(expectedNewDistribution);
+  }
+
+  private List<PartitionId> getSortedPartitionIds(final int partitionCount) {
+    return IntStream.rangeClosed(1, partitionCount)
+        .mapToObj(id -> PartitionId.from("temp", id))
+        .collect(Collectors.toList());
+  }
+
+  private Set<MemberId> getClusterMembers(final int newClusterSize) {
+    return IntStream.range(0, newClusterSize)
+        .mapToObj(Integer::toString)
+        .map(MemberId::from)
+        .collect(Collectors.toSet());
+  }
+
+  public static Stream<Arguments> provideReassignmentParameters() {
+    return Stream.of(
```

Comment:
:wrench: Let's throw in some cases with old or new cluster size = 1 and partition count = 1.
Might be a neat case for jqwik?

Reply:
Using jqwik is a nice idea. Let me try it out.

Reply:
Done.

Reply:
> Let's throw in some cases with old or new cluster size = 1 and partition count = 1.

Good suggestion. It did uncover some bugs :smile: